Atomic feature flags followup

miminar opened the pull request:
Fixes test symlinks setup and ditches speces from API constant values.

Shall be extended for product naming changes once the discussion settles down.

miminar commented:
@smarterclayton, @deads2k could you please take a look?

smarterclayton commented:
Why do we really need to follow the symlinks?  oadm and oc are probably enough for now... I'm kind of hesitant to say this is strongly important (we can just avoid using those strings in the admin commands and client tools).  Do you have a case in the client tools where we need the string?

Also, remember 'oc' is going to be delivered as a separate binary, so it's really only 'oadm' that is going to be able to use this.  Is it really important to have oadm switch?

miminar commented:
The use case is to print different help in client tools depending on running
platform. Consider current help messages:

    $ oc
    Developer and Administrator Client
    ...
    This will create an application based on the Docker image 'openshift/ruby-20-centos7' that builds
    the source code at 'github.com/openshift/ruby-hello-world.git'. To start the build, run

      $ oc start-build ruby-hello-world --follow
    ...
    Build and Deploy Commands:
      start-build  Starts a new build
    ...
      new-build    Create a new build configuration
      cancel-build Cancel a pending or running build
    ...

    $ oc describe --help
    Show details of a specific resource
    ...
      // Provide details about the ruby-sample-build build configuration
      $ oc describe bc ruby-sample-build
    ... 

It's helpful if running OpenShift but it's horse feathers on Atomic Enterprise
Platform.

Having `oc` as a separate binary is a problem in this regard. Is it a must?
Symlinking `/usr/bin/openshift` to `/usr/bin/oc` works like a charm.

smarterclayton commented:
Oc is a separate binary by request because it halves the size of the binary.  I think we want to do selective disabling of features in the oc client based on API discovery on the server in the long run, so in the short run we can live with it.

sdodson commented:
Currently the RPM packaging still symlinks openshift to oc, but I'll change that soon.

sdodson commented:
Are we still pursuing this?

danmcp commented:
@miminar Bump

miminar commented:
@sdodson I hope so.
Rebased and added support for prior space-included constants because it's quite a while since they were added and we don't want OpenShift to spit errors when parsing existing configs. Ping @smarterclayton.

liggitt commented on pkg/cmd/server/api/types.go:
```diff
@@ -8,8 +8,8 @@ import (
 	FeatureBuilder    = `Builder`
-	FeatureS2I        = `S2I Builder`
-	FeatureWebConsole = `Web Console`
+	FeatureS2I        = `S2IBuilder`
```
add a comment to this set of constants that any changes require the addition of an alias for backwards compatibility

liggitt replied:
what is driving changing these?

deads2k replied:
> what is driving changing these?

@smarterclayton made comments in the previous pull after I merged it.  He said no spaces.

liggitt replied:
ok, then do the minimal impl needed to alias these two flags, since we won't add any new flags with spaces

miminar replied:
> ok, then do the minimal impl needed to alias these two flags, since we won't add any new flags with spaces

Yeah, that's what I aimed for. More minimalistic solution would be to move and hide deprecated names to `pkg/cmd/server/api/helpers.go` and use `FeatureList` in validation code so that they don't need to be exported. But I'd rather have them visible and close to their canonical names.

liggitt commented on pkg/cmd/server/api/types.go:
```diff
@@ -20,6 +20,10 @@ var (
 	DeadKubernetesAPILevels    = []string{"v1beta1", "v1beta2"}
 	DeadOpenShiftAPILevels     = []string{"v1beta1"}
 
+	FeatureAliases = map[string][]string{
```
if you make this map aliases to their canonical feature, you get O(1) lookup and prevent duplicate aliases

liggitt replied:
also, should you normalize the alias?
```
FeatureAliases = map[string]string {
  NormalizeOpenShiftFeature("S2I Builder"): FeatureS2I,
  ...
}
```

miminar replied:
> if you make this map aliases to their canonical feature, you get O(1) lookup and prevent duplicate aliases

Good idea. Will do.

> also, should you normalize the alias?

No. `NormalizeOpenShiftFeature()` returns canonical name for any give feature flag or its alias. I'll fix description.

liggitt replied:
if you don't normalize the alias, you don't get O(1) lookup, you have to iterate. Not a big deal, since the aliases map shouldn't have more than two items in it

miminar replied:
I don't have a use case for such a lookup because of additional helper maps/sets but it might come handy. I've just made those aliases lower-cased.

sdodson commented:
@miminar / @liggitt When will we wrap this PR up?

I'm being asked for newer packages but without this change I have to create an additional set of services in order to enable openshift feature set that will be thrown out as soon as this change is merged. This will also entail ansible changes to adjust feature enablement and I'd like to get that out of the way sooner rather than later.

liggitt commented on hack/test-cmd.sh:
```diff
@@ -46,7 +46,7 @@ set -e
 function find_tests {
   cd "${OS_ROOT}"
-  find "${1}" -name '*.sh' -print0 | sort -u | xargs -0 -n1 printf "%s\n"
+  find "${1}" -name '*.sh' -print0 | sort -uz | tr '\0' '\n'
```
what is this change for?

miminar replied:
It actually sorts the cmd tests. Previous line didn't have the effect. Otherwise it's unrelated to proposed changes.

liggitt replied:
why not just `find "${1}" -name '*.sh' | sort -u`?

liggitt replied:
we already use the script name as a project name, so it can't contain spaces, etc

miminar replied:
That's fine with me.

liggitt commented on pkg/cmd/server/api/helpers.go:
```diff
@@ -26,6 +27,18 @@ func init() {
+// NormalizeOpenShiftFeature returns canonical name for given OpenShift feature
+// flag or an alias. Returns empty string if nothing matches.
+func NormalizeOpenShiftFeature(name string) string {
+	if feature, ok := knownOpenShiftFeatureSet[strings.ToLower(name)]; ok {
```
just `return knownOpenShiftFeatureSet[strings.ToLower(name)]`

miminar replied:
Ahh. That's a good one :-).

liggitt commented on pkg/cmd/server/api/helpers.go:
```diff
@@ -34,8 +47,8 @@ func (fl *FeatureList) Add(items ...string) error {
 	for _, item := range items {
-		feature, exists := knownOpenShiftFeatureSet[strings.ToLower(item)]
-		if !exists {
+		feature := NormalizeOpenShiftFeature(item)
```
I would keep the return value `(string, bool)`, checking len(...)==0 to determine validity isn't great

miminar replied:
Good point.

liggitt commented on pkg/cmd/server/api/helpers.go:
```diff
@@ -73,9 +86,9 @@ func (fl *FeatureList) Delete(items ...string) {
 // case-insensitive.
 func (fl FeatureList) Has(feature string) bool {
-	lowerCased := strings.ToLower(feature)
+	normalized := NormalizeOpenShiftFeature(feature)
```
this changes behavior from just lowercasing to limiting to the set of known flags... and also means that if the feature we're checking, and a feature in the FeatureList are both unknown, they will match (""=="")... need to rework this

miminar commented:
I have some problems running tests after a rebase. Don't merge yet.

miminar commented:
@liggitt Could you please take a look once more?

liggitt commented on pkg/cmd/openshift/openshift.go:
```diff
@@ -97,18 +97,15 @@ func CommandFor(basename string) *cobra.Command {
-		product = "Atomic"
+	distPrefix := cmdutil.GetProductName(name)
+	if distPrefix == cmdutil.ProductAtomicEnterprise {
+		distPrefix = "Atomic"
```
why are we doing this transformation here? I want this logic in one place... if we have distinct distribution, product, and platform names (which seems a little crazy), there should be three helpers, `GetDistributionName`, etc

miminar replied:
It's crazy indeed. I wonder how many products can one git repository hold...

miminar commented:
[test]
@liggitt, @deads2k, @smarterclayton WDYT?

liggitt commented on pkg/cmd/server/api/helpers.go:
```diff
@@ -26,6 +27,18 @@ func init() {
+}
+
+// NormalizeOpenShiftFeature returns canonical name for given OpenShift feature
+// flag or an alias. Returns empty string if nothing matches.
```
why return empty string for unknown features? usage would be a lot simpler with this impl:
```
name = strings.ToLower(name)
if feature, ok := knownOpenShiftFeatureSet[name]; ok {
  return feature, true
}
return name, false
```

liggitt replied:
all the places this is used, the returned bool is now checked as the signal for an unknown item, so returning the lowercased name should be fine

miminar replied:
You're right. Fixed.

liggitt replied:
fix godoc to match behavior

liggitt commented on pkg/cmd/server/api/helpers.go:
```diff
@@ -73,9 +80,13 @@ func (fl *FeatureList) Delete(items ...string) {
 // case-insensitive.
 func (fl FeatureList) Has(feature string) bool {
-	lowerCased := strings.ToLower(feature)
+	normalized, featureKnown := NormalizeOpenShiftFeature(feature)
```
if `NormalizeOpenShiftFeature` returned the lowercased name even for unknown things, this would be simpler:
```
featureNormalized, _ := NormalizeOpenShiftFeature(feature)
for _, item := range fl {
  itemNormalized, _ := NormalizeOpenShiftFeature(item)
  if itemNormalized == featureNormalized {
    return true
  }
}
```

miminar replied:
Simplified.

liggitt commented on pkg/cmd/util/product.go:
```diff
@@ -0,0 +1,50 @@
+	}
+}
+
+func GetDistributionName(binaryName string) string {
```
add godoc with an excerpt of the usage of the returned value so we don't change this to something that makes no grammatical sense laster

miminar replied:
Added.

liggitt commented on pkg/cmd/util/product.go:
```diff
@@ -0,0 +1,50 @@
+	}
+}
+
+func GetPlatformName(binaryName string) string {
```
same here, godoc with usage context

miminar commented:
@liggitt could you please give it one more look?

liggitt commented on pkg/cmd/util/product.go:
```diff
@@ -0,0 +1,55 @@
+
+// GetProductName chooses appropriate product for a binary name.
+func GetProductName(binaryName string) string {
+	name := filepath.Base(os.Args[0])
```
use the passed binaryName, not os.Args

miminar replied:
Oops. Good catch!

liggitt commented:
one doc nit, one comment about using the passed name, not os.Args, go ahead and squash

miminar commented:
@liggitt: resolved.

liggitt commented:
lgtm, [merge]
